kvdb/cache.go

```diff
@@ -0,0 +1,930 @@
+package kvdb
+
+import (
+	"container/list"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+	"github.com/google/btree"
+)
+
+const (
+	treeDeg = 3
+)
+
+type cacheBucket struct {
+	seq  *uint64
+	tree *btree.BTree
```

Comment:
So with this, we always end up caching the contents of an _entire_ bucket? I'd be concerned about the memory blow up here for larger nodes that have a large revocation log (ever growing append only log for channels), invoices, payments, etc. 

Do we have any profiles that show the resident memory overhead of this patch on mainnet?

Reply:
Yeah, this simple implementation is only allowing us to cache whole buckets from top-level down to the leafes. This is because if we want to evict whole buckets from the memory when they're inacive (LRU strategy) we'd need a more compex architecture:
- single reader/writer since both can modify the cache
- need to bubble up and create the DB read buckets if the cache needs to read in an evicted bucket (this is needed, because a huge performance bottleneck alone is just getting the bucket key/values when structure is deep).

Current PR only caches channel-state. With the `payments` bucket we really wouldn't want to read the whole bucket on start. Invoices bucket didn't seem to be very problematic in my benchmarks so far.